src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -27,32 +27,38 @@ public class HttpConnection : IConnection
 
         private readonly ILoggerFactory _loggerFactory;
         private readonly ILogger _logger;
-
         private volatile ConnectionState _connectionState = ConnectionState.Disconnected;
-        private readonly object _stateChangeLock = new object();
-
-        private volatile ChannelConnection<byte[], SendMessage> _transportChannel;
         private readonly HttpClient _httpClient;
         private readonly HttpOptions _httpOptions;
-        private volatile ITransport _transport;
-        private volatile Task _receiveLoopTask;
-        private TaskCompletionSource<object> _startTcs;
-        private TaskCompletionSource<object> _closeTcs;
-        private TaskQueue _eventQueue;
-        private readonly ITransportFactory _transportFactory;
-        private string _connectionId;
-        private Exception _abortException;
-        private readonly TimeSpan _eventQueueDrainTimeout = TimeSpan.FromSeconds(5);
-        private ChannelReader<byte[]> Input => _transportChannel.Input;
-        private ChannelWriter<SendMessage> Output => _transportChannel.Output;
         private readonly List<ReceiveCallback> _callbacks = new List<ReceiveCallback>();
         private readonly TransportType _requestedTransportType = TransportType.All;
+        private readonly ITransportFactory _transportFactory;
+        private readonly object _stateChangeLock = new object();
+
+        private HttpConnectionOneOff _connection;
+        private List<Action<Exception>> _closed = new List<Action<Exception>>();
 
         public Uri Url { get; }
 
         public IFeatureCollection Features { get; } = new FeatureCollection();
 
-        public event Action<Exception> Closed;
+        public event Action<Exception> Closed
```

Comment:
Can we not just change _closed to be a Thread Safe type?

Reply:
Accessing shouldn't be frequent, in fact only 1 access should happen at a time in our code.